src/api/java/com/minecolonies/api/util/InventoryUtils.java

```diff
@@ -2149,13 +2180,47 @@ public static boolean tryRemoveStackFromItemHandler(final IItemHandler handler,
                 }
             }
         }
-
-        final ItemStack revertStack = input.copy();
-        revertStack.setCount(input.getCount() - amount);
-        addItemStackToItemHandler(handler, revertStack);
+        // This should never happen.
         return false;
     }
 
+    /**
+     * Tries to remove a group of ItemStorages with its size from a given ItemHandler. Only removes stacks if every stack's full count is present.
+     *
+     * @param handler      the itemHandler.
+     * @param input        the List of ItemStorage to remove.
+     * @return true if removed the stacks
+     */
+    public static boolean tryRemoveStorageFromItemHandler(final IItemHandler handler, final List<ItemStorage> input)
+    {
+        for(final ItemStorage is : input)
+        {
+            if(!hasItemStackInItemHandler(handler, is.getItemStack(), is.ignoreDamageValue(), is.ignoreNBT()))
+            {
+                return false;
+            }
+        }
+
+        for(final ItemStorage is : input)
+        {
+            int amount = is.getItemStack().getCount();
+
+            for (int i = 0; i < handler.getSlots(); i++)
+            {
+                if (ItemStackUtils.compareItemStacksIgnoreStackSize(handler.getStackInSlot(i), is.getItemStack()))
+                {
+                    amount = amount - handler.extractItem(i, amount, false).getCount();
+
+                    if (amount == 0)
+                    {
+                        break;
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
```

Comment:
As far as I can tell, all of the similar functions use either require NBT and damage matching (recipes, builder items), or use Predicates (builder tools).  Because one of the possible supported (and default) items for the Undo Research function is Ancient Tomes and this was better handled in the server-config file than a datapack, we need some approach for item selection which can be implemented in the config readably.

Once it came to that, the previous method was functional enough for me to use and copy, but had potential issues and the change to this implementation was requested as part of the conversation [here](https://github.com/ldtteam/minecolonies/pull/6181#discussion_r548477514) with Raycoms.